### PR TITLE
[Refactor] New expression model

### LINKED ISSUE #14: [Refactoring] Fix `QSymbol` on parametric operators like `RX`
Currently the `QSymbol` does not support parametric addressing for operators like `RX` properly, for instance, the following example should be expected to happen:

```python
theta = Symbol("theta")
RX = QSymbol("RX")
RX(theta)()  # support all qubits
```
but in the current `__call__` method does not account for `params`, only `support` (qubits).

### PR DESCRIPTION
**Pull request by kaosmicadei**
## General
It simplifies the `Expression` implementation, reducing the number of classes/types.

`Value` (numerical values) and `Symbol` are now expressions.

The previous `QSymbol` has evolved into a regular symbol wrapped by a `QuantumOperator` expression. This change simplifies the code and allows us to define parametric quantum operators using regular functions wrapped in `QuantumOperators`.

then | now
--|--
`QSymbol("X")` | `QuantumOperator(Symbol("X"), Support(...))`
`QSymbol("RX", angle)` | `QuantumOperator(Function("RX", angle), Support(...))`

## Refactoring:
- [x] Expression class
- [x] Dagger/conjugate method
- [x] Evaluation:
    - [x] Multiplication
    - [x] Addition
    - [x] Subtraction
    - [x] Negation
    - [x] Division
    - [x] Power
    - [x] KroneckerProduct
- [x] Replace
- [x] Remove the `Numeric` protocol and `Numerical` type alias

## Fix:
- [x] Type hints

Closes #14

### CODE REVIEW

File: qadence2_expressions/core/support.py [hunk 1]
```diff
@@ -23,7 +25,7 @@ class Support:
 
     def __init__(
         self,
-        *indices: int,
+        *indices: Any,
```
[the remark this answers is not in the document]
Comment (Doomsk) [reply qadence2_expressions/core/support.py]: `args` (`*args`) in mypy does not accept anything other than `Any` as typing annotation

File: qadence2_expressions/core/ircompiler.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+    Reserve one slot for a scaler parameter in the environment and n-slots for
+    an array. The type of the parameter is defined by the backend.
+
+    Inputs:
+        size: Space occuped by the parameter.
+        trainable: Flag if the parameter can change during a training loop.
+    """
+
+    def __init__(self, size: int, trainable: bool) -> None:
+        self.size = size
```
[the remark this answers is not in the document]
Comment (kaosmicadei) [reply qadence2_expressions/core/ircompiler.py]: This one comes from Pulser

File: qadence2_expressions/core/ircompiler.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+
+    def __repr__(self) -> str:
+        params = f"{repr(self.name)}, {self.support}"
+        args = ", ".join(map(repr, self.args))
+        if args:
+            params += ", " + args
+        return f"{self.__class__.__name__}({params})"
+
+
+class AllocQubits:
```
[the remark this answers is not in the document]
Comment (kaosmicadei) [reply qadence2_expressions/core/ircompiler.py]: Classical allocation and quantum allocation are different things. Quantum allocation is related to the register.

File: qadence2_expressions/core/ircompiler.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+            defined by the backend. Default value is 1.
+        options: Extra register related properties that may not be supported by
+            all backends.
+    """
+
+    def __init__(
+        self,
+        num_qubits: int,
+        qubit_positions: list[tuple[int, int]] | list[int],
+        grid_type: Literal["linear", "square", "triangular"] | None = None,
```
[the remark this answers is not in the document]
Comment (kaosmicadei) [reply qadence2_expressions/core/ircompiler.py]: Maybe in the future, but IDEs usually use `Literal` for hinting.

**Review by Doomsk — APPROVED** [review]
seems good to go

**Pull request merged**